### ISSUE TITLE
Initialization failed without more trials

### ISSUE THREAD

**Issue by gnool** [colmap/colmap#427]
I am currently running sparse reconstruction on a set of data (confidential so I can't share it here). It always attempts to first initialize with image 10 and 12 (I specified -1 for both image_id1 and image_id2). The normal routine is that with initialization with this pair, it will never converge in the global BA during initialization, so it will proceed with another image pair for initialization, and then finally successfully reconstruct the model. Strangely, if I repeat the reconstruction, sometimes when it failed to initialize with 10 and 12, it just ends the whole process without trying other image pair for initialization. This happens in both command line and GUI.

I'm running on Ubuntu 18.04, compiled Colmap with Ceres (and also activated "use_pba" option).
init_num_trials is at default of 200, while init_max_reg_trials is at 2. Is there any reason for this odd behavior?

**Comment by gnool** [comment]
Interestingly, on the same dataset, besides the odd behavior for initialization mentioned above, another odd behavior (after repeated running) is that sometimes Colmap gets stuck permanently. htop shows that when this occurs, all my CPU threads are in maximum usage. This happens not so frequently as compared to the issue above (the one above occurs about 50% of the time), probably 1 out of 100 cases.
After looking at the output from Colmap while it gets stuck, it seems that after the failed initialization (using image 10 and 12), it did not attempt to use other images for initialization, but rather proceed to the registration stage.
While I cannot attach the images, I can provide the .db file and the settings file project.ini in the attached zip file. Using dummy images one should be able to run the same thing as I did.
Any idea why these issues are occurring?
[debug.zip](https://github.com/colmap/colmap/files/2195879/debug.zip)

**Comment by gnool** [comment]
Thanks for testing on your machine, @ahojnnes.
The images are taken using smartphone camera. I am not sure whether I understand your comment correctly, do you mean large radial distortion coefficients are to be expected, or that I have mistakenly set large radial distortion coefficients in the .db file? I have the intention of setting them to zero.
I remember using a camera model with 5 parameters, f, cx, cy, k1, k2 and disabled refine_extra_params, so k1 and k2 should always be equal to their initial values, zero. cx and cy were set to half of image width and height respectively. Only f is refined if I recall correctly. I also disabled sharing intrinsics as the camera focal length may change slightly due to auto-focus. Are you seeing otherwise on your machine based on the provided debug file?

On my machine if I manually set init image id 1 and 2 to be 10 and 14, then the above two issues don't seem to occur. The output model shape is also good. However if they are set to -1, about half of the cases will get correct output, about half of the cases will get the first issue and produces no 3d points, and about 1% of the cases will just get stuck during reconstruction.

**Comment by gnool** [comment]
I have attached another two log files from Colmap from two runs on same dataset with same settings. In log_normal.txt, we see that once the initialization with image 10 and 12 failed, it moves on to find other pairs (and continue the whole process until the whole model is successfully generated). In log_stuck.txt (from the case where it freezes forever), it tried image 10 and 12 and failed, but it starts registering image 13. Is this behavior expected?
[log_normal.txt](https://github.com/colmap/colmap/files/2197794/log_normal.txt)
[log_stuck.txt](https://github.com/colmap/colmap/files/2197795/log_stuck.txt)

**Comment by ahojnnes** [comment]
Can you send me your database?

**Comment by gnool** [comment]
Sorry for the lack of reply. I have since then compiled a non-randomized version of Colmap and so did not have the chance to test the issues above further. 

However I am still not sure whether I understand the initialization step correctly. I have attached two .db files below. In both cases I use a simple pinhole camera model (only refine focal length and nothing else) shared across all images. However the two .db files were deliberately prepared to differ in their initial focal length provided. The ground truth is about f=2400, and the two .db files have f=3000 and f=4800 as initial focal lengths. 

In the f=4800 case, I am not sure why Colmap chooses to use image 4 and 5 as initial image pair (which are taken from very similar angles and therefore inlier_config was set to 6, as evidenced by the high ratio of H/F). When Colmap uses this image pair for initialization, it just fails to generate any model in the end.

In the f=3000 case, Colmap could automatically choose a better initial image pair with higher angle of separation, and successfully reconstruct the whole model correctly.

Some of my questions include:

- Is this behavior entirely related to the accuracy of the initial focal length provided?

- Why does Colmap not try the other image pairs for initialization other than image 4 and 5?

- In the case where one cannot provide an accurate initial focal length, what should be done to avoid the problems faced by f=4800 above?

As mentioned, here are the .db files and the project.ini file.
[project_files.zip](https://github.com/colmap/colmap/files/2265544/project_files.zip)
